### PR TITLE
Node client refactor

### PR DESCRIPTION
**Pull request by jkodumal**
This modularizes the node client, changes the polling model to poll-per-second (the new style), paves the way for redis-cached feature stores, enables LDD mode, and adds support for indirect stream events.

### CODE REVIEW

File: redis_feature_store.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    client.exists('$initialized$', function(err, obj) {
+      if (!err && obj) {
+        cache.set('$initialized$', true);
+      } 
+      cb(!err && obj);
+    });
+  }
+
+  store.close = function() {
+    client.quit();
```
Comment (jkodumal) [inline redis_feature_store.js]: I'm having a problem with the test driver when the Redis feature store is configured-- I can't get the program to exit cleanly, even if I call `client.close()`. Still investigating.

Comment (jkodumal) [inline redis_feature_store.js]: This was a fun one. I spent a couple hours chasing down red herrings on the redis side of things, but it turns out that my driver program *was exiting cleanly*, but after a 30 second delay. That made me realize that the problem was the caching library, as I'd set the cache TTL to 30 seconds. So turns out that the caching library doesn't give you an option to dispose of the timer, and any program will wait for the last timer to run and flush cache entries.

I switched to an alternative caching library that has better disposal options.

File: redis_feature_store.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
+var redis = require('redis'),
+    NodeCache = require( "node-cache" );
+
+
+function RedisFeatureStore(redis_opts, cache_ttl) {
```
Comment (jkodumal) [inline redis_feature_store.js]: @apucacao this is an optional submodule, but I'm not sure what the preferred pattern is for importing things like this. Right now, my driver does it with the following:

```
var RedisFeatureStore = require('ldclient-node/redis_feature_store');
```

but it sounds like that's not idiomatic? What's the preferred way to do this?

**Comment by jkodumal** [comment]
This is now passing the integration harness in both streaming and non-streaming modes. I'm about to add a new redis / LDD node-restwrapper instance to test the `RedisFeatureStore`, but this should be reviewable now.

**Comment by jkodumal** [comment]
The node redis restwrapper is now up and running, and passes integration tests.

File: index.js [hunk 3]
```diff
@@ -1,260 +1,156 @@
...
 
     else if (!key) {
-      send_flag_event(client, key, user, default_val, default_val);
+      config.logger.error("[LaunchDarkly] No feature flag key specified. Returning default value.");
+      send_flag_event(key, user, default_val, default_val);
       cb(new Error("[LaunchDarkly] No flag key specified in toggle call"), default_val);
       return;
     }
 
     else if (!user) {
```
Comment (drichelson) [inline index.js]: We also should check here to see if the user key exists and return the default if it is not there

**Comment by drichelson** [comment]
I will definitely defer thumbs to @apucacao on this one.

File: redis_feature_store.js (hunk 2 above)
Comment (jkodumal) [inline redis_feature_store.js]: On @apucacao 's suggestion, you would do this now with:

```
var RedisFeatureStore = require(`ldclient-node`).RedisFeatureStore;
```

File: index.js [hunk 4]
```diff
@@ -1,260 +1,157 @@
...
+var PollingProcessor = require('./polling');
+var StreamingProcessor = require('./streaming');
 var pointer = require('json-pointer');
 var tunnel = require('tunnel');
+var winston = require('winston');
 var VERSION = "1.6.0";
 
 var noop = function(){};
 
 global.setImmediate = global.setImmediate || process.nextTick.bind(process);
```
Comment (apucacao) [inline index.js]: Do we still need this? I don't see it used anywhere.

**Comment by apucacao** [comment]
This looks good to me @jkodumal . 👍

**Comment by jkodumal** [comment]
Great. I'm going to merge this onto a long running branch for v2. I don't think we'll release this independently with v1, but I will keep this branch open to make that option easier if necessary.

**Pull request closed without merging**